File: php/bin/keyschedule_check.php
<?php
// Reads the shared key-schedule vector
// (spec/corpus/transcripts/keyschedule.json) and confirms this PHP symmetric
// state reproduces every output. Invoked by interop/check-keyschedule-php.sh.
require __DIR__ . '/../src/autoload.php';

use Bonemesh\KeySchedule;

$path = $argv[1] ?? null;
if ($path === null) {
    fwrite(STDERR, "usage: keyschedule_check <keyschedule.json>\n");
    exit(2);
}
$doc = json_decode(file_get_contents($path), true);
$in = $doc['inputs'];
$out = $doc['outputs'];
$failures = 0;
$check = function (string $name, string $want, string $got) use (&$failures) {
    if (bin2hex($got) === $want) {
        echo "PASS $name\n";
    } else {
        echo "FAIL $name\n  got:  " . bin2hex($got) . "\n  want: $want\n";
        $failures++;
    }
};

$s = new KeySchedule();
$check('h_init', $out['h_init'], $s->h);
$s->mixHash(hex2bin($in['mesh_hex']));
$check('h_after_mesh', $out['h_after_mesh'], $s->h);
$s->mixKey(hex2bin($in['ss_dh_hex']));
$check('ck_after_dh', $out['ck_after_dh'], $s->ck);
$s->mixKey(hex2bin($in['ss_kem_hex']));
$check('ck_after_kem', $out['ck_after_kem'], $s->ck);
$check('ct1', $out['ct1_hex'], $s->encryptAndHash(hex2bin($in['plaintext1_hex'])));
$check('h_after_ct1', $out['h_after_ct1'], $s->h);
$check('ct2', $out['ct2_hex'], $s->encryptAndHash(hex2bin($in['plaintext2_hex'])));
$check('h_after_ct2', $out['h_after_ct2'], $s->h);
[$i2r, $r2i] = $s->split();
$check('transport_key_i2r', $out['transport_key_i2r'], $i2r);
$check('transport_key_r2i', $out['transport_key_r2i'], $r2i);

if ($failures > 0) {
    fwrite(STDERR, "$failures output(s) mismatched\n");
    exit(1);
}
echo "key schedule reproduces every shared output\n";

File: python/bin/keyschedule_check.py
#!/usr/bin/env -S uv run --script
# /// script
# requires-python = ">=3.12"
# dependencies = ["bonemesh"]
#
# [tool.uv.sources]
# bonemesh = { path = "..", editable = true }
# ///
"""Reads the shared key-schedule vector (spec/corpus/transcripts/keyschedule.json)
and confirms this symmetric state reproduces every output.

Agreement with the other implementations over this file means a handshake driven
by any of them derives the same transport keys. Invoked by
interop/check-keyschedule-python.sh.
"""

import json
import sys

from bonemesh.keyschedule import KeySchedule


def main() -> int:
    if len(sys.argv) != 2:
        print("usage: keyschedule_check <keyschedule.json>", file=sys.stderr)
        return 2
    with open(sys.argv[1], encoding="utf-8") as fh:
        doc = json.load(fh)
    i, o = doc["inputs"], doc["outputs"]
    hx = bytes.fromhex
    failures = 0

    def check(name: str, got: bytes, want: str) -> None:
        nonlocal failures
        if got.hex() == want:
            print(f"PASS {name}")
        else:
            print(f"FAIL {name}\n  got:  {got.hex()}\n  want: {want}")
            failures += 1

    s = KeySchedule()
    check("h_init", s.h, o["h_init"])
    s.mix_hash(hx(i["mesh_hex"]))
    check("h_after_mesh", s.h, o["h_after_mesh"])
    s.mix_key(hx(i["ss_dh_hex"]))
    check("ck_after_dh", s.ck, o["ck_after_dh"])
    s.mix_key(hx(i["ss_kem_hex"]))
    check("ck_after_kem", s.ck, o["ck_after_kem"])
    check("ct1", s.encrypt_and_hash(hx(i["plaintext1_hex"])), o["ct1_hex"])
    check("h_after_ct1", s.h, o["h_after_ct1"])
    check("ct2", s.encrypt_and_hash(hx(i["plaintext2_hex"])), o["ct2_hex"])
    check("h_after_ct2", s.h, o["h_after_ct2"])
    i2r, r2i = s.split()
    check("transport_key_i2r", i2r, o["transport_key_i2r"])
    check("transport_key_r2i", r2i, o["transport_key_r2i"])

    if failures:
        print(f"{failures} output(s) mismatched", file=sys.stderr)
        return 1
    print("key schedule reproduces every shared output")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
